### PR TITLE
Momentum-conserving field gather

### PR DESCRIPTION
**Pull request by WeiqunZhang**
This pull request implements coarse to fine interpolation for the case that the coarse patch and fine patch have stagger grids, whereas the auxiliary data are on nodal grids.

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
Looks it doesn't work with do_nodal.  I will fix it.

**Comment by WeiqunZhang** [comment]
I think `warpx.do_nodal=1` works now.

**Comment by WeiqunZhang** [comment]
A 2d test using python failed because it had the old parameter `algo.field_gather=standard`.  Just pushed a fix.

File: Docs/source/running_cpp/parameters.rst [hunk 1]
```diff
@@ -598,7 +598,7 @@ Numerics and algorithms
      - ``direct``: simpler current deposition algorithm, described in
        the section :doc:`../theory/picsar_theory`. Note that this algorithm is not strictly charge-conserving.
 
-    If ``algo.current_deposition`` is not specified, the default is ``esirkepov``.
+    v
```
[the remark this answers is not in the document]
Comment (WeiqunZhang) [reply Docs/source/running_cpp/parameters.rst]: That's Remi's typo.  I will fix it.

File: Source/Parallelization/WarpXRegrid.cpp [hunk 2]
```diff
@@ -1,5 +1,6 @@
 
 #include <WarpX.H>
+#include <WarpXAlgorithmSelection.H>
```
[the remark this answers is not in the document]
Comment (WeiqunZhang) [reply Source/Parallelization/WarpXRegrid.cpp]: We do not need it.  I will delete it.

File: Source/WarpX.cpp [hunk 3]
```diff
@@ -853,7 +860,18 @@ WarpX::AllocLevelMFs (int lev, const BoxArray& ba, const DistributionMapping& dm
     //
     // The Aux patch (i.e., the full solution)
     //
-    if (lev == 0)
+    if (aux_is_nodal and !do_nodal)
+    {
+        BoxArray const nba = amrex::convert(ba,IntVect::TheUnitVector());
```
[the remark this answers is not in the document]
Comment (WeiqunZhang) [reply Source/WarpX.cpp]: Comment added.

File: Source/Particles/Sorting/Partition.cpp [hunk 4]
```diff
@@ -56,7 +56,7 @@ PhysicalParticleContainer::PartitionParticlesInBuffers(
         gather_masks : current_masks;
     // - For each particle, find whether it is in the larger buffer,
     //   by looking up the mask. Store the answer in `inexflag`.
-    amrex::ParallelFor( np, fillBufferFlag(pti, bmasks, inexflag, Geom(lev), 0) );
+    amrex::ParallelFor( np, fillBufferFlag(pti, bmasks, inexflag, Geom(lev)) );
```
[the remark this answers is not in the document]
Comment (WeiqunZhang) [reply Source/Particles/Sorting/Partition.cpp]: This comes from Remi's commit d0b0c0940205a7153.

Author: Remi Lehe
Date:   Fri Oct 11 16:00:03 2019 -0700

    Fix particle partition in buffers

File: Source/Particles/Sorting/Partition.cpp [hunk 5]
```diff
@@ -93,7 +93,7 @@ PhysicalParticleContainer::PartitionParticlesInBuffers(
             // - For each particle in the large buffer, find whether it is in
             // the smaller buffer, by looking up the mask. Store the answer in `inexflag`.
             amrex::ParallelFor( np - n_fine,
-               fillBufferFlag(pti, bmasks, inexflag, Geom(lev), n_fine) );
+               fillBufferFlagRemainingParticles(pti, bmasks, inexflag, Geom(lev), pid, n_fine) );
```
[the remark this answers is not in the document]
Comment (WeiqunZhang) [reply Source/Particles/Sorting/Partition.cpp]: Same

[the remark this answers is not in the document]
Comment (WeiqunZhang) [reply Source/Particles/Sorting/Partition.cpp]: Remi's changes have been merged into dev branch.

**Comment by WeiqunZhang** [comment]
I think I have addressed all the comments.

**Comment by WeiqunZhang** [comment]
I will add a regression test.

**Comment by MaxThevenet** [comment]
Thanks! I'll merge when tests pass.

**Pull request merged**